File: apps/server/src/game/MapObjects.ts
/**
 * MapObjects v10 — 맵 오브젝트 시스템
 * XP Shrine, Healing Spring, Upgrade Altar, Speed Gate
 * 아레나 내 고정 위치 상호작용 오브젝트
 */

import type {
  ArenaConfig, Position, MapObjectType, MapObjectNetworkData,
} from '@snake-arena/shared';
import { MAP_OBJECT_CONFIGS } from '@snake-arena/shared';
import type { AgentEntity } from './AgentEntity';

// ─── Internal Types ───

interface MapObject {
  id: string;
  type: MapObjectType;
  position: Position;
  radius: number;
  active: boolean;
  /** 쿨다운 종료 틱 (0 = 항상 활성) */
  cooldownUntil: number;
  /** 재사용 쿨다운 틱 수 */
  respawnTicks: number;
}

/** 맵 오브젝트 효과 적용 결과 (레벨업 이벤트 전달용) */
export interface MapObjectEvent {
  type: 'level_up';
  agentId: string;
}

// ─── MapObjects Manager ───

export class MapObjects {
  private objects: MapObject[] = [];
  /** Speed Gate 효과를 받은 에이전트 추적 (중복 적용 방지) */
  private speedGateTracker: Map<string, number> = new Map(); // agentId → cooldownUntil tick
  /** Altar 사용한 에이전트 추적 (라운드당 1회) */
  private altarUsed: Set<string> = new Set();

  constructor(arenaRadius: number) {
    this.spawnAll(arenaRadius);
  }

  /** 모든 맵 오브젝트 스폰 */
  private spawnAll(arenaRadius: number): void {
    this.objects = [];
    let idCounter = 0;

    for (const config of Object.values(MAP_OBJECT_CONFIGS)) {
      for (let i = 0; i < config.count; i++) {
        const pos = this.randomObjectPosition(arenaRadius);
        this.objects.push({
          id: `mo_${config.type}_${idCounter++}`,
          type: config.type,
          position: pos,
          radius: config.radius,
          active: true,
          cooldownUntil: 0,
          respawnTicks: config.respawnTicks,
        });
      }
    }
  }

  /** 아레나 내 랜덤 위치 생성 (중심/가장자리 회피) */
  private randomObjectPosition(arenaRadius: number): Position {
    // 중심에서 30~70% 반경 사이에 배치
    const minR = arenaRadius * 0.30;
    const maxR = arenaRadius * 0.70;

    const angle = Math.random() * Math.PI * 2;
    const r = minR + Math.random() * (maxR - minR);

    return {
      x: Math.cos(angle) * r,
      y: Math.sin(angle) * r,
    };
  }

  /**
   * 매 틱 업데이트: 에이전트 근접 체크 + 효과 적용
   * @returns 레벨업 등 Arena에서 처리해야 할 이벤트 목록
   */
  update(agents: Map<string, AgentEntity>, currentTick: number): MapObjectEvent[] {
    const events: MapObjectEvent[] = [];

    for (const obj of this.objects) {
      // 쿨다운 체크 → 재활성화
      if (!obj.active && obj.cooldownUntil > 0 && currentTick >= obj.cooldownUntil) {
        obj.active = true;
        obj.cooldownUntil = 0;
      }

      if (!obj.active) continue;

      // 에이전트 근접 체크
      for (const agent of agents.values()) {
        if (!agent.isAlive) continue;

        const dx = agent.position.x - obj.position.x;
        const dy = agent.position.y - obj.position.y;
        const distSq = dx * dx + dy * dy;
        const rSq = obj.radius * obj.radius;

        if (distSq >= rSq) continue;

        // 범위 내 → 타입별 효과 적용
        switch (obj.type) {
          case 'shrine':
            this.applyShrine(agent, obj, currentTick);
            break;
          case 'spring':
            this.applySpring(agent);
            break;
          case 'altar':
            if (this.applyAltar(agent, obj, currentTick)) {
              events.push({ type: 'level_up', agentId: agent.data.id });
            }
            break;
          case 'gate':
            this.applyGate(agent, obj, currentTick);
            break;
        }
      }
    }

    return events;
  }

  /** XP Shrine: +15 XP 즉시, 쿨다운 적용 */
  private applyShrine(agent: AgentEntity, obj: MapObject, currentTick: number): void {
    // XP 추가 (레벨업 여부는 Arena에서 addXp로 처리하므로 직접 호출)
    agent.addXp(15);
    obj.active = false;
    obj.cooldownUntil = currentTick + obj.respawnTicks;
  }

  /** Healing Spring: +5 mass/tick (항상 활성, 범위 내에 있는 동안 매틱 적용) */
  private applySpring(agent: AgentEntity): void {
    agent.addMass(5);
  }

  /**
   * Upgrade Altar: 즉시 레벨업 (1회용, 긴 쿨다운)
   * @returns 레벨업 발생 여부
   */
  private applyAltar(agent: AgentEntity, obj: MapObject, currentTick: number): boolean {
    // 이미 사용한 에이전트는 재사용 불가
    if (this.altarUsed.has(agent.data.id)) return false;

    this.altarUsed.add(agent.data.id);
    obj.active = false;
    obj.cooldownUntil = currentTick + obj.respawnTicks;

    // 강제 레벨업 (addXp 대신 직접 레벨업 트리거)
    return true;
  }

  /** Speed Gate: 3초 스피드 부스트 (통과 시) */
  private applyGate(agent: AgentEntity, obj: MapObject, currentTick: number): void {
    // 동일 에이전트 중복 적용 방지
    const lastUsed = this.speedGateTracker.get(agent.data.id);
    if (lastUsed !== undefined && currentTick < lastUsed) return;

    // 3초 (60틱) 스피드 효과 부여
    agent.addEffect('speed', 60, currentTick);
    this.speedGateTracker.set(agent.data.id, currentTick + 60);

    obj.active = false;
    obj.cooldownUntil = currentTick + obj.respawnTicks;
  }

  /** 네트워크 직렬화 데이터 반환 */
  getNetworkData(): MapObjectNetworkData[] {
    return this.objects.map(obj => ({
      id: obj.id,
      type: obj.type,
      x: Math.round(obj.position.x),
      y: Math.round(obj.position.y),
      r: obj.radius,
      active: obj.active,
    }));
  }

  /** 라운드 리셋 */
  reset(arenaRadius: number): void {
    this.objects = [];
    this.speedGateTracker.clear();
    this.altarUsed.clear();
    this.spawnAll(arenaRadius);
  }
}

File: server/internal/game/map_objects.go
package game

import (
	"fmt"
	"math"
	"math/rand"

	"github.com/andrewkim-gif/snake/server/internal/domain"
)

// MapObjectType classifies the kind of map object.
type MapObjectType string

const (
	MapObjXPShrine     MapObjectType = "shrine"
	MapObjHealingSpring MapObjectType = "spring"
	MapObjUpgradeAltar MapObjectType = "altar"
	MapObjSpeedGate    MapObjectType = "gate"
)

// MapObject represents an interactive object placed in the arena.
type MapObject struct {
	ID           string        `json:"id"`
	Type         MapObjectType `json:"type"`
	Position     domain.Position `json:"position"`
	Active       bool          `json:"active"`
	CooldownTicks int          `json:"-"` // ticks remaining until reactivation
	RespawnTicks  int          `json:"-"` // total cooldown ticks for this type
	UsesLeft     int           `json:"-"` // for Altar: uses remaining this round (-1 = unlimited)
}

// MapObjectManager manages all arena map objects.
type MapObjectManager struct {
	objects map[string]*MapObject
	config  MapObjectConfig
}

// NewMapObjectManager creates a new MapObjectManager with default config.
func NewMapObjectManager() *MapObjectManager {
	cfg := DefaultMapObjectConfig()
	return &MapObjectManager{
		objects: make(map[string]*MapObject),
		config:  cfg,
	}
}

// PlaceObjects spawns all map objects at their designated positions.
// Called at round start after arena reset.
func (m *MapObjectManager) PlaceObjects(arenaRadius float64) {
	m.objects = make(map[string]*MapObject)

	// --- XP Shrines: 3 at random positions within 60% radius ---
	for i := 0; i < m.config.ShrineCount; i++ {
		angle := rand.Float64() * 2 * math.Pi
		r := arenaRadius * 0.6 * math.Sqrt(rand.Float64())
		m.objects[fmt.Sprintf("shrine_%d", i)] = &MapObject{
			ID:           fmt.Sprintf("shrine_%d", i),
			Type:         MapObjXPShrine,
			Position:     domain.Position{X: math.Cos(angle) * r, Y: math.Sin(angle) * r},
			Active:       true,
			CooldownTicks: 0,
			RespawnTicks:  m.config.ShrineCooldownSec * TickRate,
			UsesLeft:     -1, // unlimited
		}
	}

	// --- Healing Springs: 2 at edge positions (80% radius) ---
	for i := 0; i < m.config.SpringCount; i++ {
		angle := rand.Float64() * 2 * math.Pi
		r := arenaRadius * 0.8
		m.objects[fmt.Sprintf("spring_%d", i)] = &MapObject{
			ID:           fmt.Sprintf("spring_%d", i),
			Type:         MapObjHealingSpring,
			Position:     domain.Position{X: math.Cos(angle) * r, Y: math.Sin(angle) * r},
			Active:       true,
			CooldownTicks: 0,
			RespawnTicks:  m.config.SpringCooldownSec * TickRate,
			UsesLeft:     -1,
		}
	}

	// --- Upgrade Altar: 1 at center (0,0) ---
	m.objects["altar_0"] = &MapObject{
		ID:           "altar_0",
		Type:         MapObjUpgradeAltar,
		Position:     domain.Position{X: 0, Y: 0},
		Active:       true,
		CooldownTicks: 0,
		RespawnTicks:  0, // no respawn; uses-limited per round
		UsesLeft:     m.config.AltarUsesPerRound,
	}

	// --- Speed Gates: 4 at cardinal directions (50% radius, 90 degree intervals) ---
	for i := 0; i < m.config.GateCount; i++ {
		angle := float64(i) * math.Pi / 2 // 0, 90, 180, 270 degrees
		r := arenaRadius * 0.5
		m.objects[fmt.Sprintf("gate_%d", i)] = &MapObject{
			ID:           fmt.Sprintf("gate_%d", i),
			Type:         MapObjSpeedGate,
			Position:     domain.Position{X: math.Cos(angle) * r, Y: math.Sin(angle) * r},
			Active:       true,
			CooldownTicks: 0,
			RespawnTicks:  m.config.GateCooldownSec * TickRate,
			UsesLeft:     -1,
		}
	}
}

// Update ticks all cooldowns and re-activates map objects when ready.
func (m *MapObjectManager) Update(tick uint64) {
	for _, obj := range m.objects {
		if obj.Active {
			continue
		}
		// Altar has no respawn cooldown — it is use-limited per round
		if obj.Type == MapObjUpgradeAltar {
			continue
		}
		obj.CooldownTicks--
		if obj.CooldownTicks <= 0 {
			obj.Active = true
			obj.CooldownTicks = 0
		}
	}
}

// CheckCollisions tests all active map objects against all alive agents.
// Returns a list of activation events. The caller (Arena) applies effects.
func (m *MapObjectManager) CheckCollisions(agents map[string]*domain.Agent, tick uint64) []MapObjectActivation {
	var activations []MapObjectActivation
	collectRadiusSq := MapObjectCollectRadius * MapObjectCollectRadius

	for _, obj := range m.objects {
		if !obj.Active {
			continue
		}

		for _, agent := range agents {
			if !agent.Alive {
				continue
			}
			dx := agent.Position.X - obj.Position.X
			dy := agent.Position.Y - obj.Position.Y
			if dx*dx+dy*dy <= collectRadiusSq {
				activations = append(activations, MapObjectActivation{
					Object:  obj,
					AgentID: agent.ID,
				})
				// Deactivate this object (one agent per activation)
				m.deactivate(obj)
				break
			}
		}
	}

	return activations
}

// deactivate puts a map object on cooldown.
func (m *MapObjectManager) deactivate(obj *MapObject) {
	obj.Active = false
	switch obj.Type {
	case MapObjUpgradeAltar:
		obj.UsesLeft--
		if obj.UsesLeft <= 0 {
			// Altar exhausted for this round — stays inactive
			obj.CooldownTicks = 0
		}
	default:
		obj.CooldownTicks = obj.RespawnTicks
	}
}

// GetNearbyMapObjects returns active map objects within a given radius of (x, y).
// Used by bot AI for observation.
func (m *MapObjectManager) GetNearbyMapObjects(x, y, radius float64) []*MapObject {
	radiusSq := radius * radius
	var result []*MapObject
	for _, obj := range m.objects {
		if !obj.Active {
			continue
		}
		dx := obj.Position.X - x
		dy := obj.Position.Y - y
		if dx*dx+dy*dy <= radiusSq {
			result = append(result, obj)
		}
	}
	return result
}

// GetAllObjects returns all map objects (for serialization/broadcasting).
func (m *MapObjectManager) GetAllObjects() []*MapObject {
	result := make([]*MapObject, 0, len(m.objects))
	for _, obj := range m.objects {
		result = append(result, obj)
	}
	return result
}

// Reset clears all map objects (called at round start before PlaceObjects).
func (m *MapObjectManager) Reset() {
	m.objects = make(map[string]*MapObject)
}

// MapObjectActivation describes a single agent-object interaction event.
type MapObjectActivation struct {
	Object  *MapObject
	AgentID string
}

// ApplyMapObjectEffect applies the effect of an activated map object to the agent.
// Returns a human-readable effect description string.
func ApplyMapObjectEffect(agent *domain.Agent, obj *MapObject, cfg MapObjectConfig) string {
	switch obj.Type {
	case MapObjXPShrine:
		// +50% XP buff for 10 seconds (ticks)
		duration := cfg.ShrineEffectSec * TickRate
		AddEffect(agent, "xp_shrine_buff", duration, cfg.ShrineXPMultiplier)
		return fmt.Sprintf("+%.0f%% XP for %ds", cfg.ShrineXPMultiplier*100, cfg.ShrineEffectSec)

	case MapObjHealingSpring:
		// +20% mass recovery (based on InitialMass as reference)
		healAmount := agent.Mass * cfg.SpringMassRecovery
		agent.Mass += healAmount
		return fmt.Sprintf("+%.0f mass healed", healAmount)

	case MapObjUpgradeAltar:
		// Instant level-up
		if agent.Level < domain.MaxLevel {
			PerformLevelUp(agent)
			return fmt.Sprintf("Instant level-up to Lv%d", agent.Level)
		}
		return "Already max level"

	case MapObjSpeedGate:
		// x2 speed for 5 seconds
		duration := cfg.GateEffectSec * TickRate
		AddEffect(agent, "speed_gate_buff", duration, cfg.GateSpeedMult)
		return fmt.Sprintf("x%.0f speed for %ds", cfg.GateSpeedMult, cfg.GateEffectSec)

	default:
		return ""
	}
}
